crates/key-value-aws/src/store.rs

```diff
@@ -0,0 +1,528 @@
+            region,
+            consistent_read,
+            table: Arc::new(table),
+            client: async_once_cell::Lazy::from_future(client_fut),
+        })
+    }
+}
+
+#[async_trait]
+impl StoreManager for KeyValueAwsDynamo {
+    async fn get(&self, _name: &str) -> Result<Arc<dyn Store>, Error> {
+        Ok(Arc::new(AwsDynamoStore {
+            client: self.client.get_unpin().await.clone(),
+            table: self.table.clone(),
+            consistent_read: self.consistent_read,
+        }))
+    }
+
+    fn is_defined(&self, _store_name: &str) -> bool {
+        true
+    }
+
+    fn summary(&self, _store_name: &str) -> Option<String> {
+        Some(format!(
+            "AWS DynamoDB region: {}, table: {}",
+            self.region, self.table
+        ))
+    }
+}
+
+struct AwsDynamoStore {
+    // Client wraps an Arc so should be low cost to clone
+    client: Client,
+    table: Arc<String>,
+    consistent_read: bool,
+}
+
+#[derive(Debug, Clone)]
+enum CasState {
+    // Existing item with version
+    Versioned(String),
+    // Existing item without version
+    Unversioned(Blob),
+    // Item was null when fetched during `current`
+    Unset,
+    // Potentially new item -- `current` was never called to fetch version
+    Unknown,
+}
+
+struct CompareAndSwap {
+    key: String,
+    client: Client,
+    table: Arc<String>,
+    bucket_rep: u32,
+    state: Mutex<CasState>,
+}
+
+/// Primary key in DynamoDB items used for querying items
+const PK: &str = "PK";
+/// Value key in DynamoDB items storing item value as binary
+const VAL: &str = "VAL";
+/// Version key in DynamoDB items used for atomic operations
+const VER: &str = "VER";
+
+#[async_trait]
+impl Store for AwsDynamoStore {
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Error> {
+        let response = self
+            .client
+            .get_item()
+            .consistent_read(self.consistent_read)
+            .table_name(self.table.as_str())
+            .key(
+                PK,
+                aws_sdk_dynamodb::types::AttributeValue::S(key.to_string()),
+            )
+            .projection_expression(VAL)
+            .send()
+            .await
+            .map_err(log_error)?;
+
+        let item = response.item.and_then(|mut item| {
+            if let Some(AttributeValue::B(val)) = item.remove(VAL) {
+                Some(val.into_inner())
+            } else {
+                None
+            }
+        });
+
+        Ok(item)
+    }
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error> {
+        self.client
+            .put_item()
+            .table_name(self.table.as_str())
+            .item(PK, AttributeValue::S(key.to_string()))
+            .item(VAL, AttributeValue::B(Blob::new(value)))
+            .send()
+            .await
+            .map_err(log_error)?;
+        Ok(())
+    }
+
+    async fn delete(&self, key: &str) -> Result<(), Error> {
+        if self.exists(key).await? {
```

Comment:
Why do we need to check for key existence? Does `client.delete_item` fail if the item doesn't exist? If so, we still have a race condition between the call to `exists` and `delete_item` where the item might be deleted and the error could occur.

Reply:
Good point! It appears that the delete operation only fails if a condition is set. I will remove this check as the API does not throw errors when running delete on nonexistent items